### DISCUSSION TITLE
How to restrict some keywords or create one-line katas?

### DISCUSSION THREAD

**Discussion by cestxvcdim** [codewars/codewars.com#3290] [Q&A]
Hello, I'm facing a problem that I don't know how to limit a user's decision in using keywords. How, for example, are one-line tasks created? After all, the solution must be strictly one line, often with a character limit.

**Comment by zoeparker123** [comment]
You can enforce one-line katas or keyword restrictions by adding custom validation in the test suite. Most creators use simple regex checks to block certain words or to make sure the full solution fits within a strict character or line limit. It’s not a perfect system, but it’s the standard way Codewars handles these kinds of constraints.

**Reply by maderiaya** [reply]
Hello! This is a very specific and interesting challenge, often encountered when creating coding katas (like on Codewars) or specific validation systems.
The short answer is: You generally can't (and shouldn't) try to parse or "understand" the logic of the code a user writes using regular Python. Instead, you control the environment and the interface.
Here is how you can enforce "one-liner" or "strict keyword" constraints, based on different scenarios:
1. The "Restricted Python Environment" (Most Secure)
If you are building a web app or a testing server (like an online judge), you never run the user's raw code directly. You run it in a "sandbox."
How to enforce "One Line":

**Comment by jessya020** [comment]
To restrict users to one-line, keyword-bound tasks (with a character limit), combine front-end guidance and back-end enforcement for clarity and security:
Start with the UI: Use a single-line text input (not a multi-line area) to block line breaks by default, add real-time character count feedback (e.g., "50/100 chars") that turns red when exceeding the limit, and disable the submit button until the count is valid. Add clear labels (e.g., "Single-line task (max 100 chars) – start with TODO:/FIX:") and inline error messages (e.g., "Use 'TODO:' at the start" or "No line breaks allowed") that pop up as the user types.
On the back end, even if the front end checks, re-verify: Scan for hidden line breaks (\n/\r) and reject multi-line inputs; check character length and reject over-limit content; use a regex (e.g., /^(TODO:|FIX:)\s.+) to confirm the input starts with required keywords. Return specific error responses (e.g., "Task must be one line" or "Missing 'TODO:' keyword") so users know how to fix issues. This two-layer approach keeps submissions strict while keeping the process intuitive for users.a
